src/ripple/app/ledger/impl/InboundLedgers.cpp

```diff
@@ -69,76 +70,99 @@ class InboundLedgersImp : public InboundLedgers
+                    inbound = std::make_shared<InboundLedger>(
+                        app_,
+                        hash,
+                        seq,
+                        reason,
+                        std::ref(m_clock),
+                        mPeerSetBuilder->build());
+                    mLedgers.emplace(hash, inbound);
+                    inbound->init(sl);
+                    ++mCounter;
+                }
             }
-        }
 
-        if (inbound->isFailed())
-            return {};
+            if (inbound->isFailed())
+                return {};
 
-        if (!isNew)
-            inbound->update(seq);
+            if (!isNew)
+                inbound->update(seq);
 
-        if (!inbound->isComplete())
-            return {};
+            if (!inbound->isComplete())
+                return {};
 
-        if (reason == InboundLedger::Reason::HISTORY)
-        {
-            if (inbound->getLedger()->stateMap().family().isShardBacked())
-                app_.getNodeStore().storeLedger(inbound->getLedger());
-        }
-        else if (reason == InboundLedger::Reason::SHARD)
-        {
-            auto shardStore = app_.getShardStore();
-            if (!shardStore)
+            if (reason == InboundLedger::Reason::HISTORY)
             {
-                JLOG(j_.error())
-                    << "Acquiring shard with no shard store available";
-                return {};
+                if (inbound->getLedger()->stateMap().family().isShardBacked())
+                    app_.getNodeStore().storeLedger(inbound->getLedger());
             }
-            if (inbound->getLedger()->stateMap().family().isShardBacked())
-                shardStore->setStored(inbound->getLedger());
-            else
-                shardStore->storeLedger(inbound->getLedger());
-        }
-        return inbound->getLedger();
+            else if (reason == InboundLedger::Reason::SHARD)
+            {
+                auto shardStore = app_.getShardStore();
+                if (!shardStore)
+                {
+                    JLOG(j_.error())
+                        << "Acquiring shard with no shard store available";
+                    return {};
+                }
+                if (inbound->getLedger()->stateMap().family().isShardBacked())
+                    shardStore->setStored(inbound->getLedger());
+                else
+                    shardStore->storeLedger(inbound->getLedger());
+            }
+            return inbound->getLedger();
+        };
+        using namespace std::chrono_literals;
+        std::shared_ptr<Ledger const> ledger = perf::measureDurationAndLog(
+            doAcquire, "InboundLedgersImp::acquire", 500ms, j_);
+
+        return ledger;
+    }
+
+    void
+    acquireAsync(
+        uint256 const& hash,
+        std::uint32_t seq,
+        InboundLedger::Reason reason) override
+    {
+        std::unique_lock lock(acquiresMutex_);
+        if (pendingAcquires_.contains(hash))
```

Comment:
In the context of `acquireLedger` we use ledger ID and ledger hash interchangeably. Is it possible that in the case of using ID, we might be skipping on ledger update that is different from the one that already applied if we use ID as a differentiator?
i.e. `adaptor_.acquireLedger(prevLedgerID) in Consensus.h`.

Reply:
ledger ID and ledger hash are the same thing.

Reply:
We discussed this offline with Mark and I don't have any reservations about ledger identity.